go/runfiles/manifest.go

```diff
@@ -97,10 +107,121 @@ func (m manifest) path(s string) (string, error) {
 	// prefixes of path in the manifest.
 	for prefix := s; prefix != ""; prefix, _ = path.Split(prefix) {
 		prefix = strings.TrimSuffix(prefix, "/")
-		if prefixMatch, ok := m[prefix]; ok {
+		if prefixMatch, ok := m.m[prefix]; ok {
 			return prefixMatch + filepath.FromSlash(strings.TrimPrefix(s, prefix)), nil
 		}
 	}
 
 	return "", os.ErrNotExist
 }
+
+func (m *manifest) open(name string) (fs.File, error) {
+	if name != "." {
+		r, err := m.path(name)
+		if err == ErrEmpty {
+			return emptyFile(name), nil
+		} else if err == nil {
+			return os.Open(r)
+		}
+	}
+
+	// At this point the file is not directly listed in the manifest (or
+	// contained in a directory that is). We lazily build a trie to allow
+	// efficient listing of the contents of intermediate directories.
+	m.initTrie()
+
+	dir := m.trie
+	if name != "." {
+		segments := strings.Split(name, "/")
+		for _, s := range segments {
+			if dir == nil {
+				break
+			}
+			dir = dir[s]
+		}
+		if dir == nil {
+			return nil, os.ErrNotExist
+		}
+	}
+
+	var entries []manifestDirEntry
+	for e := range dir {
+		entries = append(entries, manifestDirEntry{e, m.m[path.Join(name, e)]})
+	}
+	sort.Slice(entries, func(i, j int) bool {
+		return entries[i].name < entries[j].name
+	})
+	return &manifestReadDirFile{dirFile(path.Base(name)), entries}, nil
+}
+
+func (m *manifest) initTrie() {
+	if m.trie != nil {
+		return
+	}
+	m.trie = make(map[string]trie)
+	for k := range m.m {
```

Comment:
This will only trigger if the call to `m.path` at the beginning of the method fails, which means that either the user requested a file that doesn't exist (should be rare in passing tests) or explicitly wants to access a directory rather than a file. In this sense it is already lazy.